p2p/src/cybershake.rs

```diff
@@ -286,6 +378,121 @@ impl<W: AsyncRead + Unpin> Incoming<W> {
     }
 }
 
+impl<W: AsyncRead + Unpin> Incoming<W> {
+    fn decipher_buf(&mut self) {
+        let seq = self.seq;
+        self.seq += 1;
+
+        self.kdf.append_u64(b"seq", seq);
+        let mut key = [0u8; 32];
+        self.kdf.challenge_bytes(b"key", &mut key);
+
+        let ad = encode_u64le(seq);
+
+        let siv_tag = GenericArray::clone_from_slice(&self.buf.as_slice()[..16]);
+        Aes128PmacSiv::new(GenericArray::clone_from_slice(&key))
+            .decrypt_in_place_detached(
+                &[&ad],
+                &mut self.buf.as_mut_slice()[16..self.ciphertext_length as usize],
+                &siv_tag,
+            )
+            .unwrap();
```

Comment:
If decryption fails (normally, because auth tag is invalid), we should return a ProtocolError and not panic.

Reply:
We cannot return any other error than `io::Error` from `poll_* methods. We can return `io::Error::Custom(io::ErrorKind::InvalidData, "protocol error")` as example

Reply:
let's kill our Error and use io::Error with InvalidData kind for version mismatch and encryption failures.

Reply:
Done